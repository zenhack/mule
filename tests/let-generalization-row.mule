#XFAIL?
let f = fn x. x.y in
  { a = f { y = {} }
  , b = f { y = { z = {} } }
  }
