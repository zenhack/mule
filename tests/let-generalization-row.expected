
Desugared:

	(letrec (types ()) (values ((f (fn x (. x y)))))
	 (record (types ())
	  (values
	   ((a (f (record (types ()) (values ((y {}))))))
	    (b
	     (f
	      (record (types ())
	       (values ((y (record (types ()) (values ((z {}))))))))))))))

inferred type:

	(record (types ())
	 (values
	  ((b (record (types ()) (values ((z (record (types ()) (values ())))))))
	   (a (record (types ()) (values ()))))))

Runtime term:

	(LetRec
	 (((0 (Lam (0 () (App ((GetField y) (Var 0)))))))
	  (LetRec
	   (((3 (App ((Var 2) (LetRec (((0 (Record ()))) (Record ((y (Var 0)))))))))
	     (3
	      (App
	       ((Var 2)
	        (LetRec
	         (((0 (LetRec (((0 (Record ()))) (Record ((z (Var 0))))))))
	          (Record ((y (Var 0))))))))))
	    (Record ((a (Var 0)) (b (Var 1))))))))

Evaluated:

	(Record ((a (Record ())) (b (Record ((z (Record ())))))))
